apps/api/src/workspace-role/service/workspace-role.service.ts

```diff
@@ -204,23 +236,46 @@ export class WorkspaceRoleService {
       )
     }
 
-    if (dto.projectSlugs) {
+    if (dto.projectEnvironments) {
       await this.prisma.projectWorkspaceRoleAssociation.deleteMany({
         where: {
           roleId: workspaceRoleId
         }
       })
 
       const projectSlugToIdMap = await this.getProjectSlugToIdMap(
-        dto.projectSlugs
+        dto.projectEnvironments
+          .map((pe) => pe.projectSlug)
+          .filter((slug) => slug)
       )
 
-      await this.prisma.projectWorkspaceRoleAssociation.createMany({
-        data: dto.projectSlugs.map((projectSlug) => ({
-          roleId: workspaceRoleId,
-          projectId: projectSlugToIdMap.get(projectSlug)
-        }))
-      })
+      for (const pe of dto.projectEnvironments) {
+        const projectId = projectSlugToIdMap.get(pe.projectSlug)
+        if (projectId) {
+          // Create or Update the project workspace role association with the environments accessible on the project
+          await this.prisma.projectWorkspaceRoleAssociation.upsert({
+            where: {
+              roleId_projectId: {
+                roleId: workspaceRoleId,
+                projectId: projectId
+              }
+            },
+            update: {
+              environments: {
```

Comment:
Can you explain what this update does?

Reply:
Sets the previously connected environments to empty then adds the new ones.